shm_wrapper_aux/shm_wrapper_aux.c

```diff
@@ -0,0 +1,469 @@
+	}
+	
+	//close both semaphores
+	if (sem_close(gp_sem) == -1) {
+		(process == NET_HANDLER) ? error("sem_close: gamepad_mutex@net_handler") : error("sem_close: gamepad_mutex@client");
+	}
+	if (sem_close(rd_sem) == -1) {
+		(process == NET_HANDLER) ? error("sem_close: robot_desc_mutex@net_handler") : error("sem_close: robot_desc_mutex@client");
+	}
+	
+	//the network handler is also responsible for unlinking everything
+	if (process == NET_HANDLER) {
+		//unlink shared memory blocks
+		if (shm_unlink(GPAD_SHM_NAME) == -1) {
+			error("shm_unlink");
+		}
+		if (shm_unlink(ROBOT_DESC_SHM_NAME) == -1) {
+			error("shm_unlink");
+		}
+		
+		//unlink semaphores
+		if (sem_unlink(GP_MUTEX_NAME) == -1) {
+			error("sem_unlink: gamepad_mutex");
+		}
+		if (sem_unlink(RD_MUTEX_NAME) == -1) {
+			error("sem_unlink: robot_desc_mutex");
+		}
+	}
+}
+
+/*
+This function reads the specified field.
+Blocks on the robot description semaphore.
+	- process: one of DEV_HANDLER, EXECUTOR, NET_HANDLER to specify which process this function is
+		being called from
+	- field: one of the robot_desc_val_t's defined above to read from
+Returns one of the robot_desc_val_t's defined above that is the current value of that field.
+*/
+robot_desc_val_t robot_desc_read (process_t process, robot_desc_field_t field)
+{
+	robot_desc_val_t ret;
+	
+	//wait on rd_sem
+	if (sem_wait(rd_sem) == -1) {
+		error("sem_wait: robot_desc_mutex");
+	}
+	
+	//read the value out, and turn off the appropriate element
+	ret = rd_ptr->fields[field];
+	if (process == EXECUTOR) {
+		rd_ptr->net_to_exe_pending = 0;
+	} else if (process == NET_HANDLER) {
+		rd_ptr->exe_to_net_pending = 0;
+	}
+	
+	//release rd_sem
+	if (sem_post(rd_sem) == -1) {
+		error("sem_post: robot_desc_mutex");
+	}
+	
+	return ret;
+}
+
+
+/*
+This function writes the specified value into the specified field.
+Blocks on the robot description semaphore.
+	- process: one of DEV_HANDLER, EXECUTOR, NET_HANDLER to specify which process this function is
+		being called from
+	- field: one of the robot_desc_val_t's defined above to write val to
+	- val: one of the robot_desc_vals defined above to write to the specified field
+No return value.
+*/
+void robot_desc_write (process_t process, robot_desc_field_t field, robot_desc_val_t val)
+{
+	struct timeval start, curr;
+	double time_taken = 0.0;
+	char msg[64]; //for error message
+	
+	//loop for a maximum of DESC_WR_MAXBLK seconds before logging a warning and writing
+	gettimeofday(&start, NULL);
+	while (time_taken < DESC_WR_MAXBLK) {
+		//if previously written data was pulled out
+		if (process == NET_HANDLER && !(rd_ptr->net_to_exe_pending)) {
+			break;
+		} else if (process == EXECUTOR && !(rd_ptr->exe_to_net_pending)) {
+			break;
+		}
+		usleep(1000);
+		gettimeofday(&curr, NULL);
+		time_taken = (double)(curr.tv_sec - start.tv_sec) + ((double)(curr.tv_usec - start.tv_usec) / 1000000.0);
+	}
+	if (time_taken >= DESC_WR_MAXBLK) {
```

Comment:
I don't think it's a good idea to stall like this. I don't see any use case where this is needed, and it seems to hurt in most cases.

Reply:
So it only stalls when the previously written value hasn't been read by the receiving end yet, i.e. under normal circumstances it won't stall at all. The use case is like maybe there's some command to go into a particular state or perform some action that is sent to the robot from Dawn or Shepherd, and it gets written to that particular field (some of these might be like game-specific things, like my freshman year the game had something like this where the robot had to do a certain thing once when Shepherd gave them the command to do it). But before that command gets read by the executor, the network handler writes to that block again, overwriting the previous value, and so the robot just misses the first value completely and doesn't execute the command. It's just a safety measure to try and ensure that all the things written to the robot description block are passed on to the desired receiver. But if everything is functioning normally, it doesn't stall at all.

Reply:
In almost all cases though, we don't want to read every update. Like we don't care if the gamepad went from connected to disconnected to connected. same for dawn and shepherd states. The existing states are not important enough to read every single write from what I've seen and the network handler shouldn't be writing states extremely quickly either. Even if there was a use case I'm missing, the block is written incorrectly because it blocks every single write to all fields when it should only block the write to the specific field that changed. And same for the read flag.

Reply:
Ok I see what you're saying, like they shouldn't be writing something once and then reverting it back very quickly. I'll change this...